driver/src/main/com/mongodb/client/gridfs/GridFSBucketImpl.java

```diff
@@ -149,26 +149,16 @@ public ObjectId uploadFromStream(final String filename, final InputStream source
         int chunkSize = options.getChunkSizeBytes() == null ? chunkSizeBytes : options.getChunkSizeBytes();
         byte[] buffer = new byte[chunkSize];
         int len;
-        MongoGridFSException savedThrowable = null;
         try {
             while ((len = source.read(buffer)) != -1) {
                 uploadStream.write(buffer, 0, len);
             }
+            uploadStream.close();
         } catch (IOException e) {
-            savedThrowable = new MongoGridFSException("IO Exception when reading from the InputStream", e);
+            uploadStream.abort();
+            throw new MongoGridFSException("IOException when reading from the InputStream", e);
         } catch (Exception e) {
-            savedThrowable = new MongoGridFSException("Unexpected exception when reading stream and writing to GridFS", e);
-        } finally {
-            try {
-                uploadStream.close();
-            } catch (Exception e) {
-                if (savedThrowable == null) {
-                    throw new MongoGridFSException("Unexpected exception when closing the GridFSUploadStream", e);
-                }
-            }
-            if (savedThrowable != null) {
-                throw savedThrowable;
-            }
+            throw new MongoGridFSException("Unexpected exception when reading stream and writing to GridFS", e);
```

Comment:
OK.  But what could happen here is that either the write or close methods could throw a MongoException of some sort (or some unexpected runtime exception), and it will be wrapped in a MongoGridFSException.  So it appears to me that it would be correct to remove that final catch of Exception and just allow that exception to be raised.